File: runtime/Swift/Sources/Antlr4/atn/LL1Analyzer.swift
/// 
/// Copyright (c) 2012-2017 The ANTLR Project. All rights reserved.
/// Use of this file is governed by the BSD 3-clause license that
/// can be found in the LICENSE.txt file in the project root.
/// 


public class LL1Analyzer {
    /// 
    /// Special value added to the lookahead sets to indicate that we hit
    /// a predicate during analysis if `seeThruPreds==false`.
    /// 
    public let HIT_PRED: Int = CommonToken.INVALID_TYPE

    public let atn: ATN

    public init(_ atn: ATN) {
        self.atn = atn
    }

    /// 
    /// Calculates the SLL(1) expected lookahead set for each outgoing transition
    /// of an _org.antlr.v4.runtime.atn.ATNState_. The returned array has one element for each
    /// outgoing transition in `s`. If the closure from transition
    /// __i__ leads to a semantic predicate before matching a symbol, the
    /// element at index __i__ of the result will be `null`.
    /// 
    /// - parameter s: the ATN state
    /// - returns: the expected symbols for each outgoing transition of `s`.
    /// 
    public func getDecisionLookahead(_ s: ATNState?) -> [IntervalSet?]? {
        
        guard let s = s else {
             return nil
        }
        let length = s.getNumberOfTransitions()
        var look = [IntervalSet?](repeating: nil, count: length)
        for alt in 0..<length {
            look[alt] = IntervalSet()
            var lookBusy = Set<ATNConfig>()
            let seeThruPreds = false // fail to get lookahead upon pred
            _LOOK(s.transition(alt).target, nil, PredictionContext.EMPTY,
                    look[alt]!, &lookBusy, BitSet(), seeThruPreds, false)
            // Wipe out lookahead for this alternative if we found nothing
            // or we had a predicate when we !seeThruPreds
            if look[alt]!.size() == 0 || look[alt]!.contains(HIT_PRED) {
                look[alt] = nil
            }
        }
        return look
    }

    /// 
    /// Compute set of tokens that can follow `s` in the ATN in the
    /// specified `ctx`.
    /// 
    /// If `ctx` is `null` and the end of the rule containing
    /// `s` is reached, _org.antlr.v4.runtime.Token#EPSILON_ is added to the result set.
    /// If `ctx` is not `null` and the end of the outermost rule is
    /// reached, _org.antlr.v4.runtime.Token#EOF_ is added to the result set.
    /// 
    /// - parameter s: the ATN state
    /// - parameter ctx: the complete parser context, or `null` if the context
    /// should be ignored
    /// 
    /// - returns: The set of tokens that can follow `s` in the ATN in the
    /// specified `ctx`.
    /// 
    public func LOOK(_ s: ATNState, _ ctx: RuleContext?) -> IntervalSet {
        return LOOK(s, nil, ctx)
    }

    /// 
    /// Compute set of tokens that can follow `s` in the ATN in the
    /// specified `ctx`.
    /// 
    /// If `ctx` is `null` and the end of the rule containing
    /// `s` is reached, _org.antlr.v4.runtime.Token#EPSILON_ is added to the result set.
    /// If `ctx` is not `null` and the end of the outermost rule is
    /// reached, _org.antlr.v4.runtime.Token#EOF_ is added to the result set.
    /// 
    /// - parameter s: the ATN state
    /// - parameter stopState: the ATN state to stop at. This can be a
    /// _org.antlr.v4.runtime.atn.BlockEndState_ to detect epsilon paths through a closure.
    /// - parameter ctx: the complete parser context, or `null` if the context
    /// should be ignored
    /// 
    /// - returns: The set of tokens that can follow `s` in the ATN in the
    /// specified `ctx`.
    /// 

    public func LOOK(_ s: ATNState, _ stopState: ATNState?, _ ctx: RuleContext?) -> IntervalSet {
        let r = IntervalSet()
        let seeThruPreds = true // ignore preds; get all lookahead
        let lookContext = ctx != nil ? PredictionContext.fromRuleContext(s.atn!, ctx) : nil
        var config = Set<ATNConfig>()
        _LOOK(s, stopState, lookContext, r, &config, BitSet(), seeThruPreds, true)
        return r
    }

    /// 
    /// Compute set of tokens that can follow `s` in the ATN in the
    /// specified `ctx`.
    /// 
    /// If `ctx` is `null` and `stopState` or the end of the
    /// rule containing `s` is reached, _org.antlr.v4.runtime.Token#EPSILON_ is added to
    /// the result set. If `ctx` is not `null` and `addEOF` is
    /// `true` and `stopState` or the end of the outermost rule is
    /// reached, _org.antlr.v4.runtime.Token#EOF_ is added to the result set.
    /// 
    /// - parameter s: the ATN state.
    /// - parameter stopState: the ATN state to stop at. This can be a
    /// _org.antlr.v4.runtime.atn.BlockEndState_ to detect epsilon paths through a closure.
    /// - parameter ctx: The outer context, or `null` if the outer context should
    /// not be used.
    /// - parameter look: The result lookahead set.
    /// - parameter lookBusy: A set used for preventing epsilon closures in the ATN
    /// from causing a stack overflow. Outside code should pass
    /// `new HashSet<ATNConfig>` for this argument.
    /// - parameter calledRuleStack: A set used for preventing left recursion in the
    /// ATN from causing a stack overflow. Outside code should pass
    /// `new BitSet()` for this argument.
    /// - parameter seeThruPreds: `true` to true semantic predicates as
    /// implicitly `true` and "see through them", otherwise `false`
    /// to treat semantic predicates as opaque and add _#HIT_PRED_ to the
    /// result if one is encountered.
    /// - parameter addEOF: Add _org.antlr.v4.runtime.Token#EOF_ to the result if the end of the
    /// outermost context is reached. This parameter has no effect if `ctx`
    /// is `null`.
    /// 
    internal func _LOOK(_ s: ATNState,
                        _ stopState: ATNState?,
                        _ ctx: PredictionContext?,
                        _ look: IntervalSet,
                        _ lookBusy: inout Set<ATNConfig>,
                        _ calledRuleStack: BitSet,
                        _ seeThruPreds: Bool,
                        _ addEOF: Bool) {
        // print ("_LOOK(\(s.stateNumber), ctx=\(ctx)");
        let c = ATNConfig(s, 0, ctx)
        if lookBusy.contains(c) {
            return
        } else {
            lookBusy.insert(c)
        }

        if s == stopState {
            guard let ctx = ctx else {
                try! look.add(CommonToken.EPSILON)
                return
            }

            if ctx.isEmpty() && addEOF {
                try! look.add(CommonToken.EOF)
                return
            }

        }

        if s is RuleStopState {
            guard let ctx = ctx else {
                try! look.add(CommonToken.EPSILON)
                return
            }

            if ctx.isEmpty() && addEOF {
                try! look.add(CommonToken.EOF)
                return
            }

            if ctx != PredictionContext.EMPTY {
                // run thru all possible stack tops in ctx
                let length = ctx.size()
                for i in 0..<length {
                    let returnState = atn.states[(ctx.getReturnState(i))]!
                    let removed = try! calledRuleStack.get(returnState.ruleIndex!)
                    try! calledRuleStack.clear(returnState.ruleIndex!)
                    _LOOK(returnState, stopState, ctx.getParent(i), look, &lookBusy, calledRuleStack, seeThruPreds, addEOF)
                    if removed {
                        try! calledRuleStack.set(returnState.ruleIndex!)
                    }
                }
                return
            }
        }

        let n = s.getNumberOfTransitions()
        for i in 0..<n {
            let t = s.transition(i)
            if let rt = t as? RuleTransition {
                if try! calledRuleStack.get(rt.target.ruleIndex!) {
                    continue
                }

                let newContext = SingletonPredictionContext.create(ctx, rt.followState.stateNumber)
                try! calledRuleStack.set(rt.target.ruleIndex!)
                _LOOK(t.target, stopState, newContext, look, &lookBusy, calledRuleStack, seeThruPreds, addEOF)
                try! calledRuleStack.clear(rt.target.ruleIndex!)
            }
            else if t is AbstractPredicateTransition {
                if seeThruPreds {
                    _LOOK(t.target, stopState, ctx, look, &lookBusy, calledRuleStack, seeThruPreds, addEOF)
                } else {
                    try! look.add(HIT_PRED)
                }
            }
            else if t.isEpsilon() {
                _LOOK(t.target, stopState, ctx, look, &lookBusy, calledRuleStack, seeThruPreds, addEOF)
            }
            else if t is WildcardTransition {
                try! look.addAll(IntervalSet.of(CommonToken.MIN_USER_TOKEN_TYPE, atn.maxTokenType))
            }
            else {
                var set = t.labelIntervalSet()
                if set != nil {
                    if t is NotSetTransition {
                        set = set!.complement(IntervalSet.of(CommonToken.MIN_USER_TOKEN_TYPE, atn.maxTokenType)) as? IntervalSet
                    }
                    try! look.addAll(set)
                }
            }
        }
    }
}

File: runtime/CSharp/runtime/CSharp/Antlr4.Runtime/Atn/LL1Analyzer.cs
/* Copyright (c) 2012-2017 The ANTLR Project. All rights reserved.
 * Use of this file is governed by the BSD 3-clause license that
 * can be found in the LICENSE.txt file in the project root.
 */
using System.Collections.Generic;
using Antlr4.Runtime.Atn;
using Antlr4.Runtime.Misc;
using Antlr4.Runtime.Sharpen;

namespace Antlr4.Runtime.Atn
{
    public class LL1Analyzer
    {
        /// <summary>
        /// Special value added to the lookahead sets to indicate that we hit
        /// a predicate during analysis if
        /// <c>seeThruPreds==false</c>
        /// .
        /// </summary>
        public const int HitPred = TokenConstants.InvalidType;

        [NotNull]
        public readonly ATN atn;

        public LL1Analyzer(ATN atn)
        {
            this.atn = atn;
        }

        /// <summary>
        /// Calculates the SLL(1) expected lookahead set for each outgoing transition
        /// of an
        /// <see cref="ATNState"/>
        /// . The returned array has one element for each
        /// outgoing transition in
        /// <paramref name="s"/>
        /// . If the closure from transition
        /// <em>i</em> leads to a semantic predicate before matching a symbol, the
        /// element at index <em>i</em> of the result will be
        /// <see langword="null"/>
        /// .
        /// </summary>
        /// <param name="s">the ATN state</param>
        /// <returns>
        /// the expected symbols for each outgoing transition of
        /// <paramref name="s"/>
        /// .
        /// </returns>
        [return: Nullable]
        public virtual IntervalSet[] GetDecisionLookahead(ATNState s)
        {
            //		System.out.println("LOOK("+s.stateNumber+")");
            if (s == null)
            {
                return null;
            }
            IntervalSet[] look = new IntervalSet[s.NumberOfTransitions];
            for (int alt = 0; alt < s.NumberOfTransitions; alt++)
            {
                look[alt] = new IntervalSet();
                HashSet<ATNConfig> lookBusy = new HashSet<ATNConfig>();
                bool seeThruPreds = false;
                // fail to get lookahead upon pred
                Look(s.Transition(alt).target, null, PredictionContext.EMPTY, look[alt], lookBusy, new BitSet(), seeThruPreds, false);
                // Wipe out lookahead for this alternative if we found nothing
                // or we had a predicate when we !seeThruPreds
                if (look[alt].Count == 0 || look[alt].Contains(HitPred))
                {
                    look[alt] = null;
                }
            }
            return look;
        }

        /// <summary>
        /// Compute set of tokens that can follow
        /// <paramref name="s"/>
        /// in the ATN in the
        /// specified
        /// <paramref name="ctx"/>
        /// .
        /// <p>If
        /// <paramref name="ctx"/>
        /// is
        /// <see langword="null"/>
        /// and the end of the rule containing
        /// <paramref name="s"/>
        /// is reached,
        /// <see cref="TokenConstants.EPSILON"/>
        /// is added to the result set.
        /// If
        /// <paramref name="ctx"/>
        /// is not
        /// <see langword="null"/>
        /// and the end of the outermost rule is
        /// reached,
        /// <see cref="TokenConstants.EOF"/>
        /// is added to the result set.</p>
        /// </summary>
        /// <param name="s">the ATN state</param>
        /// <param name="ctx">
        /// the complete parser context, or
        /// <see langword="null"/>
        /// if the context
        /// should be ignored
        /// </param>
        /// <returns>
        /// The set of tokens that can follow
        /// <paramref name="s"/>
        /// in the ATN in the
        /// specified
        /// <paramref name="ctx"/>
        /// .
        /// </returns>
        [return: NotNull]
        public virtual IntervalSet Look(ATNState s, RuleContext ctx)
        {
            return Look(s, null, ctx);
        }

        /// <summary>
        /// Compute set of tokens that can follow
        /// <paramref name="s"/>
        /// in the ATN in the
        /// specified
        /// <paramref name="ctx"/>
        /// .
        /// <p>If
        /// <paramref name="ctx"/>
        /// is
        /// <see langword="null"/>
        /// and the end of the rule containing
        /// <paramref name="s"/>
        /// is reached,
        /// <see cref="TokenConstants.EPSILON"/>
        /// is added to the result set.
        /// If
        /// <paramref name="ctx"/>
        /// is not
        /// <c>PredictionContext#EMPTY_LOCAL</c>
        /// and the end of the outermost rule is
        /// reached,
        /// <see cref="TokenConstants.EOF"/>
        /// is added to the result set.</p>
        /// </summary>
        /// <param name="s">the ATN state</param>
        /// <param name="stopState">
        /// the ATN state to stop at. This can be a
        /// <see cref="BlockEndState"/>
        /// to detect epsilon paths through a closure.
        /// </param>
        /// <param name="ctx">
        /// the complete parser context, or
        /// <see langword="null"/>
        /// if the context
        /// should be ignored
        /// </param>
        /// <returns>
        /// The set of tokens that can follow
        /// <paramref name="s"/>
        /// in the ATN in the
        /// specified
        /// <paramref name="ctx"/>
        /// .
        /// </returns>
        [return: NotNull]
        public virtual IntervalSet Look(ATNState s, ATNState stopState, RuleContext ctx)
        {
            IntervalSet r = new IntervalSet();
            bool seeThruPreds = true;
			PredictionContext lookContext = ctx != null ? PredictionContext.FromRuleContext(s.atn, ctx) : null;
            Look(s, stopState, lookContext, r, new HashSet<ATNConfig>(), new BitSet(), seeThruPreds, true);
            return r;
        }

        /// <summary>
        /// Compute set of tokens that can follow
        /// <paramref name="s"/>
        /// in the ATN in the
        /// specified
        /// <paramref name="ctx"/>
        /// .
        /// <p/>
        /// If
        /// <paramref name="ctx"/>
        /// is
        /// <see cref="PredictionContext.EMPTY"/>
        /// and
        /// <paramref name="stopState"/>
        /// or the end of the rule containing
        /// <paramref name="s"/>
        /// is reached,
        /// <see cref="TokenConstants.EPSILON"/>
        /// is added to the result set. If
        /// <paramref name="ctx"/>
        /// is not
        /// <see cref="PredictionContext.EMPTY"/>
        /// and
        /// <paramref name="addEOF"/>
        /// is
        /// <see langword="true"/>
        /// and
        /// <paramref name="stopState"/>
        /// or the end of the outermost rule is reached,
        /// <see cref="TokenConstants.EOF"/>
        /// is added to the result set.
        /// </summary>
        /// <param name="s">the ATN state.</param>
        /// <param name="stopState">
        /// the ATN state to stop at. This can be a
        /// <see cref="BlockEndState"/>
        /// to detect epsilon paths through a closure.
        /// </param>
        /// <param name="ctx">
        /// The outer context, or
        /// <see cref="PredictionContext.EMPTY"/>
        /// if
        /// the outer context should not be used.
        /// </param>
        /// <param name="look">The result lookahead set.</param>
        /// <param name="lookBusy">
        /// A set used for preventing epsilon closures in the ATN
        /// from causing a stack overflow. Outside code should pass
        /// <c>new HashSet&lt;ATNConfig&gt;</c>
        /// for this argument.
        /// </param>
        /// <param name="calledRuleStack">
        /// A set used for preventing left recursion in the
        /// ATN from causing a stack overflow. Outside code should pass
        /// <c>new BitSet()</c>
        /// for this argument.
        /// </param>
        /// <param name="seeThruPreds">
        ///
        /// <see langword="true"/>
        /// to true semantic predicates as
        /// implicitly
        /// <see langword="true"/>
        /// and "see through them", otherwise
        /// <see langword="false"/>
        /// to treat semantic predicates as opaque and add
        /// <see cref="HitPred"/>
        /// to the
        /// result if one is encountered.
        /// </param>
        /// <param name="addEOF">
        /// Add
        /// <see cref="TokenConstants.EOF"/>
        /// to the result if the end of the
        /// outermost context is reached. This parameter has no effect if
        /// <paramref name="ctx"/>
        /// is
        /// <see cref="PredictionContext.EMPTY"/>
        /// .
        /// </param>
        protected internal virtual void Look(ATNState s, ATNState stopState, PredictionContext ctx, IntervalSet look, HashSet<ATNConfig> lookBusy, BitSet calledRuleStack, bool seeThruPreds, bool addEOF)
        {
            //		System.out.println("_LOOK("+s.stateNumber+", ctx="+ctx);
            ATNConfig c = new ATNConfig(s, 0, ctx);
            if (!lookBusy.Add(c))
            {
                return;
            }
            if (s == stopState)
            {
                if (ctx == null)
                {
                    look.Add(TokenConstants.EPSILON);
                    return;
                }
                else if (ctx.IsEmpty && addEOF) {
                    look.Add(TokenConstants.EOF);
                   return;
                }
            }
            if (s is RuleStopState)
            {
				if (ctx == null)
				{
					look.Add(TokenConstants.EPSILON);
					return;
				}
                else if (ctx.IsEmpty && addEOF)
                {
                    look.Add(TokenConstants.EOF);
                    return;
                }
				if (ctx != PredictionContext.EMPTY)
				{
					for (int i = 0; i < ctx.Size; i++)
					{
						ATNState returnState = atn.states[ctx.GetReturnState(i)];
						bool removed = calledRuleStack.Get(returnState.ruleIndex);
						try
						{
							calledRuleStack.Clear(returnState.ruleIndex);
							Look(returnState, stopState, ctx.GetParent(i), look, lookBusy, calledRuleStack, seeThruPreds, addEOF);
						}
						finally
						{
							if (removed)
							{
								calledRuleStack.Set(returnState.ruleIndex);
							}
						}
					}
					return;
				}
            }
            int n = s.NumberOfTransitions;
            for (int i_1 = 0; i_1 < n; i_1++)
            {
                Transition t = s.Transition(i_1);
                if (t is RuleTransition)
                {
                    RuleTransition ruleTransition = (RuleTransition)t;
                    if (calledRuleStack.Get(ruleTransition.ruleIndex))
                    {
                        continue;
                    }
                    PredictionContext newContext = SingletonPredictionContext.Create(ctx, ruleTransition.followState.stateNumber);
                    try
                    {
                        calledRuleStack.Set(ruleTransition.target.ruleIndex);
                        Look(t.target, stopState, newContext, look, lookBusy, calledRuleStack, seeThruPreds, addEOF);
                    }
                    finally
                    {
                        calledRuleStack.Clear(ruleTransition.target.ruleIndex);
                    }
                }
                else
                {
                    if (t is AbstractPredicateTransition)
                    {
                        if (seeThruPreds)
                        {
                            Look(t.target, stopState, ctx, look, lookBusy, calledRuleStack, seeThruPreds, addEOF);
                        }
                        else
                        {
                            look.Add(HitPred);
                        }
                    }
                    else
                    {
                        if (t.IsEpsilon)
                        {
                            Look(t.target, stopState, ctx, look, lookBusy, calledRuleStack, seeThruPreds, addEOF);
                        }
                        else
                        {
                            if (t is WildcardTransition)
                            {
                                look.AddAll(IntervalSet.Of(TokenConstants.MinUserTokenType, atn.maxTokenType));
                            }
                            else
                            {
                                IntervalSet set = t.Label;
                                if (set != null)
                                {
                                    if (t is NotSetTransition)
                                    {
                                        set = set.Complement(IntervalSet.Of(TokenConstants.MinUserTokenType, atn.maxTokenType));
                                    }
                                    look.AddAll(set);
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}
